File: SistemaEntregas/Core/BackendFachada.cs
using System;
using System.Collections.Generic;
using SistemaEntregas.Archivos;
using SistemaEntregas.Auth;
using SistemaEntregas.Bitacora;
using SistemaEntregas.Editor;
using SistemaEntregas.Tareas;

namespace SistemaEntregas.Core
{
    /// <summary>
    /// Fachada del sistema. Implementa el patrón Singleton para garantizar una única instancia.
    /// Centraliza el acceso a todos los subsistemas.
    /// </summary>
    public class BackendFachada
    {
        private static BackendFachada instancia;
        private static bool inicializado = false;

        // Subsistemas internos
        private readonly Dictionary<int, Usuario> usuarios = new();
        private readonly Dictionary<int, Tarea> tareas = new();
        private readonly Dictionary<int, Entrega> entregas = new();
        private readonly Dictionary<int, ArchivoP> archivos = new();
        private readonly Dictionary<int, GrupoTrabajo> grupos = new();
        private readonly Bitacora.Bitacora bitacora = new();

        private int contadorUsuarios = 1;
        private int contadorTareas = 1;
        private int contadorEntregas = 1;
        private int contadorArchivos = 1;
        private int contadorGrupos = 1;

        /// <summary>
        /// Constructor privado — patrón Singleton.
        /// </summary>
        private BackendFachada()
        {
        }

        public static BackendFachada GetInstancia()
        {
            if (instancia == null)
            {
                instancia = new BackendFachada();
                inicializado = true;
            }
            return instancia;
        }

        // ── Auth ──────────────────────────────────────────────

        public string Login(string correo, string contrasena)
        {
            foreach (var usuario in usuarios.Values)
            {
                if (usuario.IniciarSesion(correo, contrasena))
                    return $"TOKEN_{usuario.Id}_{usuario.Rol}";
            }
            return string.Empty;
        }

        public int Registrar(string correo, string nombre, string contrasena, string rol)
        {
            UsuarioFactory factory = rol.ToUpper() == "PROFESOR"
                ? new ProfesorFactory()
                : new EstudianteFactory();

            var usuario = factory.CrearUsuario(correo, nombre, contrasena);
            var id = contadorUsuarios++;
            usuarios[id] = usuario;
            return id;
        }

        // ── Tareas ────────────────────────────────────────────

        public List<Tarea> GetTareas(int usuarioId, string rol)
        {
            return new List<Tarea>(tareas.Values);
        }

        public int CrearTarea(string nombre, string descripcion, DateTime fechaEntrega, bool esGrupal, int profesorId)
        {
            var tarea = new Tarea
            {
                Id = contadorTareas,
                Nombre = nombre,
                Descripcion = descripcion,
                FechaEntrega = fechaEntrega,
                EsGrupal = esGrupal,
                FechaCreacion = DateTime.Now
            };
            tareas[contadorTareas] = tarea;
            return contadorTareas++;
        }

        public void EditarTarea(int idTarea, Tarea datos)
        {
            if (tareas.TryGetValue(idTarea, out var tarea))
            {
                tarea.Nombre = datos.Nombre;
                tarea.Descripcion = datos.Descripcion;
                tarea.FechaEntrega = datos.FechaEntrega;
                tarea.EsGrupal = datos.EsGrupal;
            }
        }

        public void EliminarTarea(int idTarea)
        {
            tareas.Remove(idTarea);
        }

        // ── Entregas ──────────────────────────────────────────

        public List<Entrega> GetEntregas(int idTarea)
        {
            return tareas.TryGetValue(idTarea, out var tarea)
                ? tarea.ObtenerEntregas()
                : new List<Entrega>();
        }

        public void RegistrarEntrega(Entrega entrega)
        {
            var id = contadorEntregas++;
            entrega.Id = id;
            entregas[id] = entrega;
            entrega.Registrar();
            bitacora.Registrar(entrega);
        }

        public void AsignarNota(int idEntrega, float nota)
        {
            if (entregas.TryGetValue(idEntrega, out var entrega))
                entrega.Nota = nota;
        }

        // ── Archivos ──────────────────────────────────────────

        public int CrearArchivo(int estudianteId, string nombre, string contenido)
        {
            var archivo = new ArchivoP
            {
                Id = contadorArchivos,
                Nombre = nombre,
                Contenido = contenido,
                FechaCreacion = DateTime.Now
            };
            archivos[contadorArchivos] = archivo;
            return contadorArchivos++;
        }

        public ResultadoEjecucion EjecutarArchivo(int idArchivo)
        {
            if (archivos.TryGetValue(idArchivo, out var archivo))
            {
                var terminal = new Terminal();
                return terminal.Ejecutar(archivo);
            }
            return new ResultadoEjecucion { Errores = "Archivo no encontrado", CodigoSalida = -1 };
        }

        // ── Grupos ────────────────────────────────────────────

        public int CrearGrupo(int idTarea, string nombre, List<int> miembros)
        {
            if (!tareas.TryGetValue(idTarea, out var tarea))
                return -1;

            var grupo = new GrupoTrabajo
            {
                Id = contadorGrupos,
                Nombre = nombre,
                Tarea = tarea
            };

            foreach (var idEstudiante in miembros)
            {
                if (usuarios.TryGetValue(idEstudiante, out var u) && u is Estudiante estudiante)
                    grupo.AgregarMiembro(estudiante);
            }

            grupos[contadorGrupos] = grupo;
            return contadorGrupos++;
        }

        // ── Bitácora ──────────────────────────────────────────

        public List<RegistroEntrega> GetHistorial(int idTarea, int idEstudiante)
        {
            if (!tareas.TryGetValue(idTarea, out var tarea)) return new List<RegistroEntrega>();
            if (!usuarios.TryGetValue(idEstudiante, out var u) || u is not Estudiante estudiante) return new List<RegistroEntrega>();

            return bitacora.ObtenerHistorial(tarea, estudiante);
        }
    }
}

File: model/Core/BackendFachada.php
<?php

namespace model\Core;

use DateTime;
use model\Auth\Estudiante;
use model\Auth\EstudianteFactory;
use model\Auth\ProfesorFactory;
use model\Auth\Usuario;
use model\Auth\UsuarioFactory;
use model\Archivos\ArchivoP;
use model\Bitacora\Bitacora;
use model\Bitacora\RegistroEntrega;
use model\Editor\Terminal;
use model\Tareas\Entrega;
use model\Tareas\GrupoTrabajo;
use model\Tareas\ResultadoEjecucion;
use model\Tareas\Tarea;

/**
 * Fachada del sistema. Implementa el patrón Singleton para garantizar una única instancia.
 * Centraliza el acceso a todos los subsistemas.
 */
class BackendFachada
{
    private static ?BackendFachada $instancia = null;
    private static bool $inicializado = false;

    // Subsistemas internos
    private array $usuarios  = [];
    private array $tareas    = [];
    private array $entregas  = [];
    private array $archivos  = [];
    private array $grupos    = [];
    private Bitacora $bitacora;

    private int $contadorUsuarios  = 1;
    private int $contadorTareas    = 1;
    private int $contadorEntregas  = 1;
    private int $contadorArchivos  = 1;
    private int $contadorGrupos    = 1;

    /**
     * Constructor privado — patrón Singleton.
     */
    private function __construct()
    {
        $this->bitacora = new Bitacora();
    }

    // Prevenir clonación e instanciación por unserialize
    private function __clone() {}
    public function __wakeup() {}

    public static function getInstancia(): self
    {
        if (self::$instancia === null) {
            self::$instancia  = new self();
            self::$inicializado = true;
        }
        return self::$instancia;
    }

    // ── Auth ──────────────────────────────────────────────

    public function login(string $correo, string $contrasena): string
    {
        foreach ($this->usuarios as $usuario) {
            if ($usuario->iniciarSesion($correo, $contrasena)) {
                return "TOKEN_{$usuario->getId()}_{$usuario->getRol()->name}";
            }
        }
        return '';
    }

    public function registrar(string $correo, string $nombre, string $contrasena, string $rol): int
    {
        $factory = strtoupper($rol) === 'PROFESOR'
            ? new ProfesorFactory()
            : new EstudianteFactory();

        $usuario = $factory->crearUsuario($correo, $nombre, $contrasena);
        $id = $this->contadorUsuarios++;
        $this->usuarios[$id] = $usuario;
        return $id;
    }

    // ── Tareas ────────────────────────────────────────────

    public function getTareas(int $usuarioId, string $rol): array
    {
        return array_values($this->tareas);
    }

    public function crearTarea(
        string $nombre,
        string $descripcion,
        DateTime $fechaEntrega,
        bool $esGrupal,
        int $profesorId
    ): int {
        $tarea = new Tarea();
        $tarea->setId($this->contadorTareas);
        $tarea->setNombre($nombre);
        $tarea->setDescripcion($descripcion);
        $tarea->setFechaEntrega($fechaEntrega);
        $tarea->setEsGrupal($esGrupal);
        $tarea->setFechaCreacion(new DateTime());

        $this->tareas[$this->contadorTareas] = $tarea;
        return $this->contadorTareas++;
    }

    public function editarTarea(int $idTarea, Tarea $datos): void
    {
        if (isset($this->tareas[$idTarea])) {
            $tarea = $this->tareas[$idTarea];
            $tarea->setNombre($datos->getNombre());
            $tarea->setDescripcion($datos->getDescripcion());
            $tarea->setFechaEntrega($datos->getFechaEntrega());
            $tarea->setEsGrupal($datos->getEsGrupal());
        }
    }

    public function eliminarTarea(int $idTarea): void
    {
        unset($this->tareas[$idTarea]);
    }

    // ── Entregas ──────────────────────────────────────────

    public function getEntregas(int $idTarea): array
    {
        return isset($this->tareas[$idTarea])
            ? $this->tareas[$idTarea]->obtenerEntregas()
            : [];
    }

    public function registrarEntrega(Entrega $entrega): void
    {
        $id = $this->contadorEntregas++;
        $entrega->setId($id);
        $this->entregas[$id] = $entrega;
        $entrega->registrar();
        $this->bitacora->registrar($entrega);
    }

    public function asignarNota(int $idEntrega, float $nota): void
    {
        if (isset($this->entregas[$idEntrega])) {
            $this->entregas[$idEntrega]->setNota($nota);
        }
    }

    // ── Archivos ──────────────────────────────────────────

    public function crearArchivo(int $estudianteId, string $nombre, string $contenido): int
    {
        $archivo = new ArchivoP();
        $archivo->setId($this->contadorArchivos);
        $archivo->setNombre($nombre);
        $archivo->setContenido($contenido);
        $archivo->setFechaCreacion(new DateTime());

        $this->archivos[$this->contadorArchivos] = $archivo;
        return $this->contadorArchivos++;
    }

    public function ejecutarArchivo(int $idArchivo): ResultadoEjecucion
    {
        if (isset($this->archivos[$idArchivo])) {
            $terminal = new Terminal();
            return $terminal->ejecutar($this->archivos[$idArchivo]);
        }

        $resultado = new ResultadoEjecucion();
        $resultado->setErrores('Archivo no encontrado');
        $resultado->setCodigoSalida(-1);
        return $resultado;
    }

    // ── Grupos ────────────────────────────────────────────

    public function crearGrupo(int $idTarea, string $nombre, array $miembros): int
    {
        if (!isset($this->tareas[$idTarea])) {
            return -1;
        }

        $grupo = new GrupoTrabajo();
        $grupo->setId($this->contadorGrupos);
        $grupo->setNombre($nombre);
        $grupo->setTarea($this->tareas[$idTarea]);

        foreach ($miembros as $idEstudiante) {
            if (isset($this->usuarios[$idEstudiante]) && $this->usuarios[$idEstudiante] instanceof Estudiante) {
                $grupo->agregarMiembro($this->usuarios[$idEstudiante]);
            }
        }

        $this->grupos[$this->contadorGrupos] = $grupo;
        return $this->contadorGrupos++;
    }

    // ── Bitácora ──────────────────────────────────────────

    public function getHistorial(int $idTarea, int $idEstudiante): array
    {
        if (!isset($this->tareas[$idTarea])) return [];
        if (!isset($this->usuarios[$idEstudiante]) || !($this->usuarios[$idEstudiante] instanceof Estudiante)) return [];

        return $this->bitacora->obtenerHistorial(
            $this->tareas[$idTarea],
            $this->usuarios[$idEstudiante]
        );
    }
}
// Los puntos nuevos en esta traducción:
// 
// - **Singleton**: En PHP se refuerza bloqueando también `__clone()` y `__wakeup()` para evitar que se creen instancias adicionales por clonación o deserialización, lo cual no es necesario en C# pero es una buena práctica en PHP.
// - **`Dictionary<K,V>`**: Se reemplaza con `array` asociativo de PHP, usando los mismos IDs enteros como claves.
// - **`TryGetValue()`**: Se sustituye con `isset()` para verificar existencia, equivalente directo y más idiomático en PHP.
// - **`u is not Estudiante`**: El pattern matching de C# se traduce como `!($u instanceof Estudiante)` en PHP.
// - **`usuario.Rol`**: Como `Rol` es un enum de PHP 8.1, se accede a su nombre como string mediante la propiedad `->name`.
// - **`tareas.Remove()`**: Se traduce como `unset()`, su equivalente directo en PHP.